### ISSUE TITLE
torch.ops.fsdp.set_  on input doesn't actually modify the input (under torch.compile)

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#132197]
```py
import torch
import torch.distributed._composable.fsdp._fsdp_param

@torch.compile(backend="aot_eager")
def f(x, l):
    z = x.sin()
    y = x + 1
    torch.ops.fsdp.set_.default(x, y)
    z2 = x.sin()
    return z2, l ** 2

x = torch.randn(3)
x_clone = x.clone()
l = torch.randn(3, requires_grad=True)

result, _ = f(x, l)
assert torch.allclose(x, x_clone + 1)
```

The assertion triggers with torch.compile but there is no problem when commenting out the torch.compile decorator

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @kwen2501 @chauhang @ezyang @penguinwu @bdhirsh

**Comment by zou3519** [comment]
cc @yf225 @bdhirsh

**Comment by bdhirsh** [comment]
hmm let me take a look the generated AOT graph doesn't look obviously wrong:
```
 ===== Forward graph 0 =====
 /home/hirsheybar/local/b/pytorch/torch/fx/_lazy_graph_module.py class GraphModule(torch.nn.Module):
    def forward(self, primals_1: "f32[3][1]cpu", primals_2: "f32[3][1]cpu"):
         # File: /home/hirsheybar/local/b/pytorch/tmp.py:7 in f, code: y = x + 1
        add: "f32[3][1]cpu" = torch.ops.aten.add.Tensor(primals_1, 1)

         # File: /home/hirsheybar/local/b/pytorch/tmp.py:8 in f, code: torch.ops.fsdp.set_.default(x, y)
        set_ = torch.ops.fsdp.set_.default(primals_1, add);  add = None

         # File: /home/hirsheybar/local/b/pytorch/tmp.py:9 in f, code: z2 = x.sin()
        sin_1: "f32[3][1]cpu" = torch.ops.aten.sin.default(primals_1);  primals_1 = None

         # File: /home/hirsheybar/local/b/pytorch/tmp.py:10 in f, code: return z2, l ** 2
        pow_1: "f32[3][1]cpu" = torch.ops.aten.pow.Tensor_Scalar(primals_2, 2)
        return (sin_1, pow_1, primals_2)
```

**Comment by bdhirsh** [comment]
Fix here: https://github.com/pytorch/pytorch/pull/132322

Fun fact: we have logic in AOTDispatcher to `detach()` forward graph inputs when:
(1) we are compiling a training graph (at least one of the graph inputs needs gradients to be computed)
(2) we can statically determine that a given input will never get gradients computed for it, even though it requires grad.

detach() is a problem though, if you are detaching an input that has a metadata mutation that is applied inside the graph (because we will apply the metadata mutation to the detached tensor, not the original input).

The fix was to convey to functionalization that `fsdp.set_` mutates metadata.
